snippets/algorithm/maximum_disjoint_subarray_sum.cpp

```diff
@@ -0,0 +1,56 @@
+vector<int> s1, s2;
+// pairs of (total_sum, positive_part)
+vector<pair<int,int>> s1_sums;
+vector<pair<int,int>> s2_sums;
+
+
+// a state is (index, positive part, negative part)
+// so the sum is positive part - negative part
+void find_all_s1_sums(int i, int pos, int neg) {
+    s1_sums.push_back({pos-neg, pos});
+    if(i == s1.size()) return;
+    find_all_s1_sums(i+1, pos, neg);
+    find_all_s1_sums(i+1, pos+s1[i], neg);
+    find_all_s1_sums(i+1, pos, neg+s1[i]);
+}
+
+void find_all_s2_sums(int i, int pos, int neg) {
+    s2_sums.push_back({pos-neg, pos});
+    if(i == s2.size()) return;
+    find_all_s2_sums(i+1, pos, neg);
+    find_all_s2_sums(i+1, pos+s2[i], neg);
+    find_all_s2_sums(i+1, pos, neg+s2[i]);
+}
+
+int maximum_disjoint_subset_sum(vector<int> &A) {
+    s1.clear();
+    s2.clear();
+    s1_sums.clear();
+    s2_sums.clear();
+    int n = A.size();
+    for(int i=0;i<n;i++) {
+        if(i<n/2) s1.push_back(A[i]);
+        else s2.push_back(A[i]);
+    }
```

Comment:
optionally, instead of explicitly splitting the array in two and adding half the elements into one new vector and the other half into another, can we just pass implicitly use the existing vector we passed in? And then our find all sums method, we also pass in the size of the subarray we're considering

Reply:
"we also pass in the size of the subarray we're considering"
don't we also need to pass the start index of the subarray we're considering?

Reply:
ah nope nvm we do not

Reply:
kattis judging server seems to be down ATM - I will push changes after I get an accepted